File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:

        result = list()

        def dfs(nums: list[int], index: int) -> None:
            if index == len(nums):
                result.append(nums[:])
                return
            for i in range(index, len(nums)):
                nums[index], nums[i] = nums[i], nums[index]
                dfs(nums, index + 1)
                nums[index], nums[i] = nums[i], nums[index]
        
        dfs(nums, 0)
        return result
        

File: Data Structures & Algorithms/subsets-ii/submission-2.java
class Solution {

    private List<List<Integer>> result = new ArrayList<>();
    private List<Integer> subset = new ArrayList<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        dfs(0, subset, nums);
        return result;
    }

    private void dfs(int index, List<Integer> subset, int[] nums) {
        result.add(new ArrayList<>(subset));

        for(int j = index; j < nums.length; j++) {
            if(j > index && nums[j] == nums[j - 1]) {continue;}
            subset.add(nums[j]);
            dfs(j + 1, subset, nums);
            subset.removeLast();
        }
    }
}
